### PR TITLE
[DSM] DDP-7941: Refactoring and kit barcode check

### PR DESCRIPTION
**Pull request by gmakharat**
## Context

Ticket: https://broadinstitute.atlassian.net/browse/DDP-7941

Problem in the ticket itself was easy to implement, but since there are upcoming tickets and bugs related to kit related feature it was good idea to do refactoring.

**Things have done:**

- Converted `KitStatusChangeRoute` as abstract base class for 4 separated routes: `KitFinalScanRoute`, `KitTrackingScanRoute`, `SentKitRoute`, `ReceivedKitsRoute`.  `KitStatusChangeRoute` gathers all common data extraction and checks for those 4 route.
- Created base abstract class called `BaseKitUseCase`, which extends `Supplier<List<ScanError>>` interface to provide common interface for different kit related feature use cases.
- Created concrete kit use case classes: `KitTrackingScanUseCase`, `KitFinalSentBaseUseCase` (Final and Sent scan features share common functionalities thus created base abstract class for them), `KitReceivedUseCase`.
- Created `ScanPayload` interface with 3 abstract methods: `getKitLabel`, `getTrackingReturnId` and `getDdpLabel`(to eliminate `leftValue` and `rightValue` properties for more clarification) since those 3 values are mainly used in kit functionalities, created concrete classes of `ScanPayload`, for each concrete route and use cases.
- Created integration tests for use case classes.

**How to run tests:**

https://user-images.githubusercontent.com/75846598/186374115-c09b87d6-b699-4d2c-8918-8e35f3e62679.mov

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

**Review by ScalaChamp — DISMISSED** [review]

**Comment by SimoneMaiwald** [comment]
add steps (for other engineers) to the PR summary about how to reproduce/test

**Review by SimoneMaiwald — CHANGES_REQUESTED** [review]
PR description is missing steps how to test the feature on the frontend side!
Final Scan page, Tracking Scan page…

**Comment by gmakharat** [comment]
> PR description is missing steps how to test the feature on the frontend side! Final Scan page, Tracking Scan page…

@SimoneMaiwald it is backend changes, frontend code needs to be tweaked to test it manually.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalScanUseCase.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    public KitFinalScanUseCase(KitPayload kitPayload, KitDao kitDao) {
+        super(kitPayload, kitDao);
+    }
+
+    @Override
+    protected Optional<ScanError> process(ScanPayload scanPayload) {
+        Optional<ScanError> result;
+        String kitLabel = scanPayload.getKitLabel();
+        String ddpLabel = scanPayload.getDdpLabel();
+        if (kitLabel.length() < 14) {
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalScanUseCase.java]: only for saliva kits do we need to check the size:
``` if(!kitDao.isBloodKit(ddpLabel))```

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalScanUseCase.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            return Optional.of(new ScanError(ddpLabel, "Barcode contains less than 14 digits, "
+                    + "You can manually enter any missing digits above."));
+        }
+        if (kitDao.isBloodKit(ddpLabel)) {
+            if (kitDao.hasTrackingScan(kitLabel)) {
+                result = updateKitRequest(kitLabel, ddpLabel);
+                trigerEventsIfSuccessfulKitUpdate(result, ddpLabel, getKitRequestShipping(kitLabel, ddpLabel));
+                KitRequestDao kitRequestDao = new KitRequestDao();
+                kitRequestDao.getKitRequestByLabel(ddpLabel).ifPresent(this::writeSampleSentToES);
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitFinalScanUseCase.java]: could we already gotten that information by combining this call with the isBloodKit query?
I don't like the 3 calls to the db to get the information for 1 scan pair

**Review by SimoneMaiwald — DISMISSED** [review]

**Review by SimoneMaiwald — APPROVED** [review]

**Pull request merged**